source/auth_credentials.c

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "auth.h"
+
+#include "io.h"
+
+#include <aws/auth/credentials.h>
+#include <aws/common/string.h>
+
+static const char *s_capsule_name_credentials_provider = "aws_credentials_provider";
+
+/**
+ * Binds a Python CredentialsProvider to a native aws_credentials_provider.
+ */
+struct credentials_provider_binding {
+    struct aws_credentials_provider *native;
+
+    /* Dependencies that must outlive this.
+     * Note that different types of providers have different dependencies */
+    PyObject *bootstrap;
+};
+
+/* Runs when the GC destroys the capsule containing the binding */
+static void s_credentials_provider_capsule_destructor(PyObject *capsule) {
+    struct credentials_provider_binding *provider = PyCapsule_GetPointer(capsule, s_capsule_name_credentials_provider);
+
+    /* Note that destructor might run due to setup failing, and some/all members might still be NULL. */
+
+    if (provider->native) {
+        aws_credentials_provider_release(provider->native);
+    }
+
+    Py_XDECREF(provider->bootstrap);
+    aws_mem_release(aws_py_get_allocator(), provider);
+}
+
+struct aws_credentials_provider *aws_py_get_credentials_provider(PyObject *credentials_provider) {
+    struct aws_credentials_provider *native = NULL;
+
+    PyObject *capsule = PyObject_GetAttrString(credentials_provider, "_binding");
+    if (capsule) {
+        struct credentials_provider_binding *binding =
+            PyCapsule_GetPointer(capsule, s_capsule_name_credentials_provider);
+        if (binding) {
+            native = binding->native;
+            AWS_FATAL_ASSERT(native);
+        }
+        Py_DECREF(capsule);
+    }
+
+    return native;
+}
+
+int s_aws_string_to_cstr_and_ssize(const struct aws_string *source, const char **out_cstr, Py_ssize_t *out_ssize) {
+    *out_cstr = NULL;
+    *out_ssize = 0;
+    if (source) {
+        if (source->len > PY_SSIZE_T_MAX) {
+            return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+        }
+        *out_cstr = (const char *)source->bytes;
```

Comment:
don't we have functions for this?

Reply:
**LOOK WHAT YOU MADE ME DO**
https://github.com/awslabs/aws-c-common/pull/534